### PR TITLE
feat: Zoom to pinch support in 360 Images

### PR DESCRIPTION
**Pull request by astrid-kg**
#### Type of change

Feat  

#### Jira ticket :blue_book:

https://cognitedata.atlassian.net/browse/REV-692

## Description :pencil:

We now track up and down pointer events and start pinch zoom when two down events are present. Pinch checks distance between latest move events and changes FOV the distance change in screen size %.

## How has this been tested? :mag:

Only on my phone which has a 147.7 x 68.7 x 8.5 mm screen with 1440 x 2960 pixels. I added `StationaryCameraManager` to Reveal Examples, so I have not really been testing 360 Images, only input in this camera mode.

It would be nice to test this on other screen sizes to see if this part makes sense:

```
    const screenSize = Math.sqrt(width * width + height * height);
    if (screenSize > 0) {
      this.setFOV(this._camera.fov - (distanceDelta * 100) / screenSize);
    }
```

I also tested on PC to check that regular zoom still functions as normal.

## Test instructions :information_source:

You can do like me and expose StationaryCameraManager to examples. But the best would be if you are able to test [this model](https://cognitedata.slack.com/archives/G0115JQ8JM8/p1669380585627129) with actual 360 images.

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [x] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [ ] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts [hunk 1]
```diff
@@ -157,15 +164,33 @@ export class StationaryCameraManager implements CameraManager {
...
-  private readonly disableDragging = (_: PointerEvent) => {
+  private disableDragging(_: PointerEvent) {
     this._isDragging = false;
+  }
+
+  private readonly onPointerUp = (event: PointerEvent) => {
+    const eventIndex = this._downEventCache.findIndex(downEvent => event.pointerId === downEvent.pointerId);
+    if (eventIndex > -1) {
+      this._downEventCache.splice(eventIndex, 1);
+    }
```
Comment (christjt) [inline viewer/packages/camera-manager/src/StationaryCameraManager.ts:179]: I believe lodash has a utility method for this. See: https://lodash.com/docs/4.17.15#pull

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts [hunk 2]
```diff
@@ -181,7 +206,40 @@ export class StationaryCameraManager implements CameraManager {
...
+  }
+
+  private pinchCamera(moveEvent: PointerEvent) {
+    if (this._downEventCache.length < 2) {
+      return;
+    }
+
+    const getEuclideanDistance = (eventOne: PointerEvent, eventTwo: PointerEvent): number => {
+      const dx = eventOne.clientX - eventTwo.clientX,
+        dy = eventOne.clientY - eventTwo.clientY;
```
Comment (christjt) [inline viewer/packages/camera-manager/src/StationaryCameraManager.ts:218]:
```suggestion
      const dx = eventOne.clientX - eventTwo.clientX;
      const dy = eventOne.clientY - eventTwo.clientY;
```

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts [hunk 3]
```diff
@@ -181,7 +206,40 @@ export class StationaryCameraManager implements CameraManager {
...
+    const sensitivityThreshold = 1.0;
+    if (Math.abs(distanceDelta) < sensitivityThreshold) {
+      this._downEventCache[downEventIndex] = preMoveEvent;
+      return;
+    }
+
+    const { width, height } = this._domElement.getBoundingClientRect();
+    const screenSize = Math.sqrt(width * width + height * height);
+    if (screenSize > 0) {
+      this.setFOV(this._camera.fov - (distanceDelta * 100) / screenSize);
```
Comment (christjt) [inline viewer/packages/camera-manager/src/StationaryCameraManager.ts:240]: Pull out the 100 in a constant with a descriptive name.

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts [hunk 4]
```diff
@@ -181,7 +206,40 @@ export class StationaryCameraManager implements CameraManager {
...
+        dy = eventOne.clientY - eventTwo.clientY;
+      return Math.sqrt(dx * dx + dy * dy);
+    };
+
+    const preMoveDistance = getEuclideanDistance(this._downEventCache[0], this._downEventCache[1]);
+    const downEventIndex = this._downEventCache.findIndex(event => event.pointerId === moveEvent.pointerId);
+    const preMoveEvent = this._downEventCache[downEventIndex];
+    this._downEventCache[downEventIndex] = moveEvent;
+    const postMoveDistance = getEuclideanDistance(this._downEventCache[0], this._downEventCache[1]);
+    const distanceDelta = postMoveDistance - preMoveDistance;
```
Comment (christjt) [inline viewer/packages/camera-manager/src/StationaryCameraManager.ts:227]: Concider moving this out (or parts of it) into a separate function. A little hard to follow what is going on.

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts [hunk 5]
```diff
@@ -27,6 +27,7 @@ export class StationaryCameraManager implements CameraManager {
   private readonly _minFOV: number;
   private readonly _stopEventTrigger: DebouncedCameraStopEventTrigger;
   private _isDragging = false;
+  private _downEventCache: Array<PointerEvent> = [];
```
Comment (christjt) [inline viewer/packages/camera-manager/src/StationaryCameraManager.ts:30]: Concider using a pointerId -> Vector2 (clientX, clientY). Would simplify the usage a little since you would not have to scan the list all the time. Also you could use built in functionality for length between two points.

**Review by christjt — CHANGES_REQUESTED** [review]
Nice feature! Just a few comments related to code hygiene.

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts (hunk 1 above)
Comment (christjt) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts:179]: Note that this will be irrelevant if you change to a map which is suggested in a comment below.

Comment (astrid-kg) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts]: Replaced this with _.remove 👍

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts (hunk 2 above)
Comment (astrid-kg) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts]: Done 💯

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts (hunk 3 above)
Comment (astrid-kg) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts]: This is just normal percentage calculation, but I pulled the entire calculation out to make this clear.

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts (hunk 4 above)
Comment (astrid-kg) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts]: Added calculateDownEventDistance(). 👑

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts (hunk 5 above)
Comment (astrid-kg) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts:31]: No. Well, I did try with the Map and although some things become simpler I struggle to find the direction of the zoom as I was depending on the order of the down events. Also, we now technically support zooming and drag at the same time, which I was not able to do. So I kinda like my current solution better.

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts [hunk 6]
```diff
@@ -181,7 +206,32 @@ export class StationaryCameraManager implements CameraManager {
...
+      this._downEventCache[indexOfMoveEvent] = preMoveDownEvent;
+      return;
+    }
+
+    const { width, height } = this._domElement.getBoundingClientRect();
+    const screenSize = Math.sqrt(width * width + height * height);
+    if (screenSize > 0) {
+      const percentage = (distanceDelta * 100) / screenSize;
+      this.setFOV(this._camera.fov + percentage);
+    }
```
Comment (christjt) [inline viewer/packages/camera-manager/src/StationaryCameraManager.ts:233]: Nit
```suggestion
    if (screenSize <= 0) {
      return;
    }
    const percentage = (distanceDelta * 100) / screenSize;
    this.setFOV(this._camera.fov + percentage);
```

**Review by christjt — APPROVED** [review]
Just a tiny nit, but other than that it looks great :)

File: viewer/packages/camera-manager/src/StationaryCameraManager.ts (hunk 6 above)
Comment (astrid-kg) [reply viewer/packages/camera-manager/src/StationaryCameraManager.ts]: Done 👍
